BasicJcloudsLocationCustomizer is an EntityInitializer

sjcorbett opened the pull request:
Repeats https://github.com/apache/brooklyn-server/pull/168 but puts the location customizers into the entity's provisioning properties rather than setting it on the entity directly. It required moving `SoftwareProcess.PROVISIONING_PROPERTIES` to `BrooklynConfigKeys` so that it could be referenced from `brooklyn-locations-jclouds`.

neykov commented on locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicJcloudsLocationCustomizer.java:
```diff
@@ -18,19 +18,72 @@
+        if (existingCustomizers == null) {
+            merged = ImmutableList.<JcloudsLocationCustomizer>of(this);
+        } else {
+            merged = Lists.newArrayListWithCapacity(1 + existingCustomizers.size());
```
Personal preference: I'd use `ImmutableList.builder()` instead (to avoid playing with sizes).

neykov replied:
I was referring to the code that follows as well. What I meant was:
```
merged = ImmutableList.builder()
                      .addAll(existingCustomizers)
                      .add(this)
                      .build();
```

neykov commented on locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/BasicJcloudsLocationCustomizerTest.java:
```diff
@@ -0,0 +1,81 @@
+        assertEquals(customizers.size(), 2, "expected two customizers in " + Iterables.toString(customizers));
+        assertTrue(customizers.contains(testCustomizer), "expected to find testCustomizer in " + Iterables.toString(customizers));
+    }
+
```
Would be useful to test `addInitializer` on already managed entities.

neykov replied:
Also testing `provisioning.properties` inheritance - for example attaching to a cluster.

sjcorbett replied:
How can you add initialisers to already managed entities? Either I'm missing something obvious or it's not possible.

neykov replied:
Thought that there's an `addInitializer` method on the entity, but that's not the case.
To add an initializer to an existing entity create the initializer and call `apply` on the entity.

neykov commented:
LGTM overall. Would like to see better initializer config integration.

sjcorbett commented:
@neykov thanks for your comments. I've removed `params` and added a test that confirms that the initialiser is applied to cluster members. Can you clarify what you meant when you said that it would be useful to test `addInitializer`?

neykov commented:
Would be useful to test applying the initializer to an existing entity (before it's started obviously). To add an initializer to an entity just call its `apply` method on the entity.
The difference is subtle - initializers in the spec are applied on the entity before they are managed. Applying it after having a managed entity could lead to different results.
That's minor though, feel free to ignore.

geomacy commented on locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/BasicJcloudsLocationCustomizer.java:
```diff
@@ -18,19 +18,73 @@
+
+    @Override
+    public void apply(EntityLocal entity) {
+        ConfigKey<Object> subkey = BrooklynConfigKeys.PROVISIONING_PROPERTIES.subKey(JcloudsLocationConfig.JCLOUDS_LOCATION_CUSTOMIZERS.getName());
```
very minor - maybe call this `customizers` rather than just `subkey`, to highlight its role

neykov commented:
Looks good, merging.
